parsers/javascript/CharIterator.js

```diff
@@ -0,0 +1,29 @@
+export default class CharIterator {
+  #chars;
+  #index;
+
+  constructor(input) {
+    this.#chars = Array.from(input);
+    this.#index = -1;
+  }
+
+  next() {
+    this.#index++;
+    const done = this.#index === this.#chars.length;
+    return done ? {
+      done,
+      value: null,
+    } : {
+      done,
+      value: this.#chars[this.#index],
+    };
+  }
+
+  peek(numberOfChars) {
+    return this.#chars.slice(this.#index + 1, this.#index + 1 + numberOfChars);
+  }
+
+  peekBack() {
```

Comment:
It is possible to implement the parser without lookbehind, and might be conceptually simpler to do it this way.

If helpful, our Swift implementation might be useful prior art:

- Tape (roughly equivalent to this iterator) https://github.com/subconsciousnetwork/subconscious/blob/main/xcode/Subconscious/Shared/Parsers/Tape.swift
- Parser https://github.com/subconsciousnetwork/subconscious/blob/main/xcode/Subconscious/Shared/Parsers/Subtext.swift

Reply:
I have removed the `peekBack()` method in favour of a simple variable to remember the previous char. I'm not sure if that fulfills the requirement of no lookbehind though.